source/matrix_free/task_info.cc

```diff
@@ -775,88 +768,42 @@ namespace internal
+      const bool                       categories_are_hp,
       const std::vector<unsigned int> &cell_vectorization_categories,
       const bool                       cell_vectorization_categories_strict,
+      const std::vector<unsigned int> &parent_relation,
       std::vector<unsigned int> &      renumbering,
       std::vector<unsigned char> &     incompletely_filled_vectorization)
     {
-      const unsigned int n_macro_cells =
-        (n_active_cells + vectorization_length - 1) / vectorization_length;
-      const unsigned int n_ghost_slots =
-        (n_ghost_cells + vectorization_length - 1) / vectorization_length;
-      const unsigned int n_boundary_cells = boundary_cells.size();
-
-      incompletely_filled_vectorization.resize(n_macro_cells + n_ghost_slots);
-      renumbering.resize(n_active_cells + n_ghost_cells,
-                         numbers::invalid_unsigned_int);
-
-      // Define the outer number of partitions. In the MPI case, we have three
-      // partitions (part before comm, part with comm, part after comm)
-      if (n_procs == 1)
-        partition_row_index.resize(3);
-      else
-        partition_row_index.resize(5);
-
-      // Initially mark the cells according to the MPI ranking
-      std::vector<unsigned char> cell_marked(n_active_cells + n_ghost_cells, 0);
-      if (n_procs > 1)
-        {
-          for (unsigned int i = 0; i < n_boundary_cells; ++i)
-            cell_marked[boundary_cells[i]] = 2;
-
-          Assert(boundary_cells.size() % vectorization_length == 0 ||
-                   boundary_cells.size() == n_active_cells,
-                 ExcInternalError());
-
-          const unsigned int n_second_slot =
-            ((n_active_cells - n_boundary_cells) / 2 / vectorization_length) *
-            vectorization_length;
-          unsigned int count = 0;
-          unsigned int c     = 0;
-          for (; c < n_active_cells && count < n_second_slot; ++c)
-            if (cell_marked[c] == 0)
-              {
-                cell_marked[c] = 1;
-                ++count;
-              }
-          for (; c < n_active_cells; ++c)
-            if (cell_marked[c] == 0)
-              cell_marked[c] = 3;
-          for (; c < n_active_cells + n_ghost_cells; ++c)
-            if (cell_marked[c] == 0)
-              cell_marked[c] = 4;
-        }
-      else
-        std::fill(cell_marked.begin(), cell_marked.end(), 1);
-
-      for (const unsigned char marker : cell_marked)
-        {
-          (void)marker;
-          Assert(marker != 0, ExcInternalError());
-        }
-
+      // Give the compiler a chance to detect that vectorization_length is a
+      // power of two, which allows it to replace integer divisions by shifts
+      unsigned int vectorization_length_bits = 0;
+      unsigned int my_length                 = vectorization_length;
+      while (my_length >>= 1)
+        ++vectorization_length_bits;
+      const unsigned int n_lanes = 1 << vectorization_length_bits;
```

Comment:
I guess you are doing this here since the class is not templated with `VectorizedArrayType`. Wouldn't it be an option to make these lines of codes to some kind of utility function? It might be used in other classes where the length is not at compile time.

Reply:
Yes, the reason is that we do not have a template and do not want it just for that parameter. We could make it a utility function (that is always inlined). I do not have time now to look for other uses - do you want to open an issue?
If someone else reads this: an integer division takes  32 clock cycles of latency and 8 cycles, whereas a right shift 1-3 clock cycles. At least smart compilers can profit from this.